python/ray/data/_internal/planner/planner.py

```diff
@@ -120,6 +120,7 @@ def _plan(self, logical_op: LogicalOperator) -> PhysicalOperator:
         for op_type, plan_fn in PLAN_LOGICAL_OP_FNS:
             if isinstance(logical_op, op_type):
                 physical_op = plan_fn(logical_op, physical_children)
+                physical_op.set_logical_operator(logical_op)
```

Comment:
How does this work for logical operators that create multiple physical operators (e.g., `Read` which creates an `InputDataBuffer` and a `TaskPoolOperator`)? Would `PhysicalOperator._logical_operator` be `None`  for everything but the last physical operator created by a single logical operator?

Reply:
the `Read` -> `InputDataBuffer->TaskPoolMapOperator` was the only case i could come up with where one `LogicalOperator` produces multiple `PhysicalOperator`s. so that's why i'm wondering if it's worth the extra change.

Reply:
actually, i can just manually call `set_logical_operator()` in `plan_read_op`, so no need to add it as a constructor arg.

Reply:
another approach is that you iterate the DAG can set logical operator for those don't have it.

Reply:
if the physical DAG is already generated, how would we know which logical operator to set?

for example (ignoring operator fusion), if we have `Read -> Map`, which gets translated to `InputDataBufer -> TaskPoolMapOperator_1 -> TaskPoolMapOperator_2, and the mapping of physical -> logical operator is:

```
{
    Read: TaskPoolMapOperator_1,
    Map: TaskPoolMapOperator_2
}
```

how would we know that the `InputDataBuffer` also originated from the `Read` op, unless we explicitly set it during the translation from logical to physical op?

Reply:
This `_plan` function generates the DAG in a post-order traverse order. 
So at this line, all nodes without logical operators must be created by the current logical operator.